homeassistant/components/spokestack/tts.py

```diff
@@ -0,0 +1,82 @@
+"""Support for the Spokestack speech service."""
+import logging
+
+from spokestack.tts.clients.spokestack import TextToSpeechClient, TTSError
+
+from homeassistant.components.tts import CONF_LANG, Provider
+from homeassistant.core import HomeAssistant
+
+from .const import (
+    CONF_KEY_ID,
+    CONF_KEY_SECRET,
+    CONF_MODE,
+    CONF_PROFILE,
+    CONF_VOICE,
+    DEFAULT_LANG,
+    DEFAULT_MODE,
+    DEFAULT_PROFILE,
+    DEFAULT_VOICE,
+    SUPPORTED_LANGUAGES,
+    SUPPORTED_MODES,
+    SUPPORTED_OPTIONS,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def get_engine(hass, config, discovery_info=None):
+    """Set up Spokestack TTS provider."""
+    return SpokestackProvider(hass)
+
+
+class SpokestackProvider(Provider):
+    """Spokestack TTS Provider."""
+
+    def __init__(self, hass: HomeAssistant):
+        """Initialize Spokestack client."""
+        self._client = TextToSpeechClient(
+            key_id=hass.data.get(CONF_KEY_ID, " "),
```

Comment:
`hass.data` is a generic dictionary shared among all integrations. `hass.data[DOMAIN]` is where you store your information. This code can never work…

Reply:
ok, so all of our configuration data should be under `hass.data["spokestack"]`? This makes sense, but I just wanted to make sure I understand.

Reply:
All the info should be in your config entry. And your config entry setup forwards it and makes it available as discocvery_info